pkg/provisioning/import.go

```diff
@@ -0,0 +1,712 @@
+// Copyright © 2023 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package provisioning
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/conduitio/conduit/pkg/connector"
+	"github.com/conduitio/conduit/pkg/foundation/cerrors"
+	"github.com/conduitio/conduit/pkg/pipeline"
+	"github.com/conduitio/conduit/pkg/processor"
+	"github.com/conduitio/conduit/pkg/provisioning/config"
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+)
+
+// Import takes a pipeline config and imports it into Conduit.
+func (s *Service) Import(ctx context.Context, newConfig config.Pipeline) error {
+	oldConfig, err := s.Export(ctx, newConfig.ID)
+	if err != nil && !cerrors.Is(err, pipeline.ErrInstanceNotFound) {
+		s.logger.Warn(ctx).Err(err).Msgf("could not export pipeline with ID %v, trying to provision new pipeline despite the error", newConfig.ID)
+		// TODO does this behavior make sense? Should we rather return the error
+		//  and skip provisioning?
+	}
+
+	var actions []action
+	actions = append(actions, s.cascadingActionsDeleteOld(oldConfig, newConfig)...)
+	actions = append(actions, s.cascadingActionsUpsertNew(oldConfig, newConfig)...)
+
+	lastActionIndex, err := s.executeActions(ctx, actions)
```

Comment:
The variable name `lastActionIndex` and what `executeActions` returns are not quite aligned. `executeActions` returns the number of actions, whereas the index would mean it's one less.

Reply:
Well if the error is not `nil` the index will point to the failing action. I could rename it to `failedActionIndex` if that sounds better?

Reply:
Yup, it does!